### PR TITLE
DSL for migrations

### PR DESCRIPTION
**Pull request by drujensen**
This adds a new DSL for migrations.  It uses the same syntax as Rails migrations. 

```
require "granite/migration"

class CreatePost < Granite::Migration
  connection :pg

  def up
    create_table :post do |t|
      t.primary :id
      t.reference :user_id
      t.string :title
      t.text :body
      t.boolean :published
      t.integer :age
      t.timestamp :published_at
      t.custom "json_field JSON"
      t.custom "uuid_field UUID"

      t.timestamps
    end

    rename_table :post, :posts

    create_index :posts, :user_id

    add_column :posts, :test, :string
    rename_column :posts, :test, :new_test
    remove_column :posts, :new_test

    execute <<-SQL
      ALTER TABLE posts ADD CONSTRAINT posts-user_id-fk FOREIGN KEY (user_id) REFERENCES users(id)
    SQL
  end

  def down
    drop_table :posts
    drop_index :posts, :user_id
  end
end
```

To execute:
```
require "granite/adapter/pg"

Granite::Connections << Granite::Adapter::Pg.new(name: "pg", url: "postgres://url")

create_post = CreatePost.new
puts create_post.generate_sql(:up)
puts create_post.generate_sql(:down)
```

### CODE REVIEW

File: src/granite/table.cr [hunk 1]
```diff
@@ -51,12 +51,12 @@ module Granite::Tables
   end
 
   macro table(name)
-    @[Granite::Table(name: {{(name.is_a?(StringLiteral) ? name : name.stringify) || nil}})]
+    @[Granite::Table(name: {{(name.is_a?(StringLiteral) ? name : name.id.stringify) || nil}})]
```
Comment (drujensen) [inline src/granite/table.cr:54]: @Blacksmoke16 by adding the `.id`, this adds support for symbols so you can do `table :posts`.

**Comment by Blacksmoke16** [comment]
Given the result of https://github.com/amberframework/granite/issues/182, is this something we actually want to include?

If so, I think I would rather remove, or at least deprecate, the current internal migrator so we don't end up with just another way to do the same thing, as this would be the third way of doing migrations.

**Comment by drujensen** [comment]
@Blacksmoke16  Over the years, I have collected feedback on why developers prefer other ORM's over granite and the lack of a DSL for DDL rises to the top.  

This is not a full featured DSL yet.  There are several options that still need to be built into it.  Also, there are some missing types yet to be added.

Currently this is a self contained class with little changes to existing code.  I could see merging or having `Migrator` use `Migration` to avoid duplicate code.

Let me know your thoughts.  I can close this if you think its not worth the effort.

**Comment by Blacksmoke16** [comment]
@drujensen Personally I'm fine with raw SQL migrations but if this is a commonly requested feature I don't have a problem with the addition.  That is, assuming you/someone will be maintaining it.

**Comment by drujensen** [comment]
@Blacksmoke16 @robacarp thanks for spending the time looking at this.  

@robacarp I agree about the `default` and there are 4 other options that rails supports.  I will look at adding them.

Regarding the specific data types per db, I agree with this as well.  I will look into refactoring out the `Fields` class to be a base class and inherit for each database.  That makes more sense.

File: spec/granite/migration/migration_spec.cr [hunk 2]
```diff
@@ -0,0 +1,147 @@
+require "../../spec_helper"
+{% begin %}
+  {% adapter_literal = env("CURRENT_ADAPTER").id %}
+  class AllFields < Granite::Migration
+    connection :{{ adapter_literal }}
+
+    def up
+      create_table :posts do |t|
+        t.primary :id
+        t.serial :serial1
```
Comment (eliasjpr) [inline spec/granite/migration/migration_spec.cr:10]: I really like the dsl, how can we specify other column parameters?

**Comment by drujensen** [comment]
Thinking about this more, this should be included in the micrate project instead of here.  Closing this PR.

**Pull request closed without merging**